core/src/main/java/org/apache/gravitino/listener/ModelEventDispatcher.java

```diff
@@ -0,0 +1,386 @@
+import org.apache.gravitino.listener.api.event.LinkModelVersionPreEvent;
+import org.apache.gravitino.listener.api.event.ListModelPreEvent;
+import org.apache.gravitino.listener.api.event.ListModelVersionPreEvent;
+import org.apache.gravitino.listener.api.event.RegisterModelPreEvent;
+import org.apache.gravitino.listener.api.info.ModelInfo;
+import org.apache.gravitino.listener.api.info.ModelVersionInfo;
+import org.apache.gravitino.model.Model;
+import org.apache.gravitino.model.ModelCatalog;
+import org.apache.gravitino.model.ModelVersion;
+import org.apache.gravitino.utils.PrincipalUtils;
+
+/**
+ * {@code ModelEventDispatcher} is a decorator for {@link ModelDispatcher} that not only delegates
+ * model operations to the underlying catalog dispatcher but also dispatches corresponding events to
+ * an {@link org.apache.gravitino.listener.EventBus} after each operation is completed. This allows
+ * for event-driven workflows or monitoring of model operations.
+ */
+public class ModelEventDispatcher implements ModelDispatcher {
+  private final EventBus eventBus;
+  private final ModelDispatcher dispatcher;
+
+  /**
+   * Constructs a {@link ModelEventDispatcher} with a specified EventBus and {@link
+   * ModelDispatcher}.
+   *
+   * @param eventBus The EventBus to which events will be dispatched.
+   * @param dispatcher The underlying {@link ModelDispatcher} that will perform the actual model
+   *     operations.
+   */
+  public ModelEventDispatcher(EventBus eventBus, ModelDispatcher dispatcher) {
+    this.eventBus = eventBus;
+    this.dispatcher = dispatcher;
+  }
+
+  /**
+   * Register a model in the catalog if the model is not existed, otherwise the {@link
+   * ModelAlreadyExistsException} will be thrown. The {@link Model} object will be created when the
+   * model is registered, users can call {@link ModelCatalog#linkModelVersion(NameIdentifier,
+   * String, String[], String, Map)} to link the model version to the registered {@link Model}.
+   *
+   * @param ident The name identifier of the model.
+   * @param comment The comment of the model. The comment is optional and can be null.
+   * @param properties The properties of the model. The properties are optional and can be null or
+   *     empty.
+   * @return The registered model object.
+   * @throws NoSuchSchemaException If the schema does not exist.
+   * @throws ModelAlreadyExistsException If the model already registered.
+   */
+  @Override
+  public Model registerModel(NameIdentifier ident, String comment, Map<String, String> properties)
+      throws NoSuchSchemaException, ModelAlreadyExistsException {
+    ModelInfo registerRequest = new ModelInfo(ident.name(), properties, comment);
+    eventBus.dispatchEvent(
+        new RegisterModelPreEvent(PrincipalUtils.getCurrentUserName(), ident, registerRequest));
+    try {
+      Model model = dispatcher.registerModel(ident, comment, properties);
+      // TODO: ModelEvent
```

Comment:
In my view, this is not a good practice for raising PRs.
Each PR is supposed to be:

- small, e.g. <500 lines of code changes;
- self-contained, with all irrelevant changes stripped off;
- good enough, i.e. not half-baked;
- easy to roll-back, if some bugs are introduced unfortunately

It is perfectly okay to raise 10+ smaller PRs a day, because we know that small steps forward make big progress.

For this specific PR, I'd suggest we implement Model Event dispatcher with test cases;
then we raise PRs for each specific events, each with its own test cases.

WDYT?

Reply:
Generally, it's preferred to make the PR small, but for the event listener PRs, most codes are similar and cohesive. I'm afraid it may be hard to propose the PR and review to split smaller PRs.

Reply:
Whatever ... if we believe this code is good enough, not half-baked, I'm fine with it.